fpga_interchange/arch_pack_clusters.cc

```diff
@@ -0,0 +1,496 @@
+            IdString driver_port(port.first);
+
+            CellInfo *driver_cell = next_cell->ports.at(driver_port).net->driver.cell;
+
+            bool cell_ok = false;
+            for (auto &cell : cluster_driver_cells.at(cluster)) {
+                if (IdString(cell) == driver_cell->type) {
+                    cell_ok = true;
+                    break;
+                }
+            }
+
+            for (auto driver_in_port : driver_cell->ports) {
+                if (cell_ok)
+                    break;
+
+                if (driver_in_port.second.type != PORT_IN)
+                    continue;
+
+                CellInfo *driver_cell_driver = driver_in_port.second.net->driver.cell;
+
+                for (auto &cell : cluster_driver_in_out_cells.at(cluster)) {
+                    if (IdString(cell) == driver_cell->type && driver_cell_driver == next_cell) {
+                        cell_ok = true;
+                        break;
+                    }
+                }
+            }
+
+            if (!cell_ok or driver_cell == nullptr)
+                continue;
+
+            BelRange bels_in_tile = getBelsByTile(next_loc.x, next_loc.y);
+            for (auto bel : bels_in_tile) {
+                auto &bel_data = bel_info(ctx->chip_info, bel);
+
+                if (bel_data.site != next_bel_data.site)
+                    continue;
+
+                IdString bel_name(bel_data.name);
+
+                auto res_placed = std::find(placed_bels.begin(), placed_bels.end(), bel_name);
+                auto res_bels = std::find(port.second.begin(), port.second.end(), bel_name);
+                bool bel_valid = ctx->isValidBelForCellType(driver_cell->type, bel);
+
+                if (!bel_valid || res_bels == port.second.end() || res_placed != placed_bels.end())
+                    continue;
+
+                //log_info("\tNet name: %s - Cell name: %s - Cell type: %s - Bel: %s\n", next_cell->ports.at(driver_port).net->name.c_str(ctx), driver_cell->name.c_str(ctx), driver_cell->type.c_str(ctx), ctx->nameOfBel(bel));
+                placement.push_back(std::make_pair(driver_cell, bel));
+                placed_bels.push_back(bel_name);
+
+                break;
+            }
+        }
+
+        // Get pattern to find the next cell in the chain
+        auto cell_pattern = cell_pattern_map.find(next_cell->name);
+        std::pair<IdString, IdString> pattern_cfg = cell_pattern->second;
+        IdString port = pattern_cfg.first;
+        if (port == IdString()) {
+            log_error("Cluster misconfiguration. None of the patterns match the '%s' cell\n", next_cell->type.c_str(ctx));
+        }
+
+        next_net = next_cell->ports[port].net;
+
+        // Check if we have reached the end of the chain
+        if (next_net == nullptr || next_net->users.size() == 0) {
+            break;
+        }
+
+        next_cell = next_net->users.at(0).cell;
+
+        next_bel = find_next_chained_bel(ctx, next_bel, ctx->id("CO3"));
```

Comment:
this hardcoded `CO3` needs either to be fixed, or a `FIXME` comment left explaining why it's there

Reply:
this was actually a leftover, this information now comes from the chip info.